paderbox/io/new_subdir.py

```diff
@@ -0,0 +1,341 @@
+             - When mkdir is True every process will get another folder.
+               i.e. each process has a folder just for this process.
+             - When mkdir is False it is likely that all get the same folder,
+               but this is not guarantied. Never use mpi with consider_mpi and
+               mkdir disabled.
+        dry_run:
+            Similar to mkdir. When true, disable mkdir and print the folder
+            name.
+
+    Returns:
+        pathlib.Path of the new subdir
+
+
+    >>> get_new_subdir('/', dry_run=True)  # root folder usually contain no digits
+    dry_run: "os.mkdir(/1)"
+    PosixPath('/1')
+
+    >>> import numpy as np
+    >>> np.random.seed(0)  # This is for doctest. Never use it in practise.
+    >>> get_new_subdir('/', id_naming=NameGenerator(), dry_run=True)
+    dry_run: "os.mkdir(/helpful_tomato_finch)"
+    PosixPath('/helpful_tomato_finch')
+    """
+
+    if consider_mpi:
+        import dlp_mpi
+        if dlp_mpi.IS_MASTER:
+            pass
+        else:
+            new_folder = None
+            new_folder = dlp_mpi.bcast(new_folder)
+            return new_folder
+
+    basedir = Path(basedir).expanduser().resolve()
+    if not basedir.exists():
+        if dry_run:
+            print(f'dry_run: "os.makedirs({basedir})"')
+            # ToDo: Make this working.
+            #       Will fail when calling os.listdir
+        else:
+            basedir.mkdir(parents=True)
+
+    if Path('/net') in basedir.parents:
+        # If nt filesystem, assert not in /net/home
+        assert Path('/net/home') not in basedir.parents, basedir
+
+    prefix_ = f'{prefix}_' if prefix else ''
+    _suffix = f'_{suffix}' if suffix else ''
+
+    for i in range(200):
+        if id_naming == 'index':
+            if prefix is None and suffix is None:
+                dir_nrs = [
+                    int(d)
+                    for d in os.listdir(str(basedir))
+                    if (basedir / d).is_dir() and d.isdigit()
+                ]
+                _id = max(dir_nrs + [0]) + 1
+            else:
+                def remove_pre_suf(d):
+                    return _removesuffix(
+                        _removeprefix(str(d), prefix_),
+                        _suffix
+                    )
+
+                dir_nrs = [
+                    int(remove_pre_suf(d))
+                    for d in os.listdir(str(basedir))
+                    if (basedir / d).is_dir()
+                    if fnmatch.fnmatch(d, f'{prefix_}*{_suffix}')
+                    if remove_pre_suf(d).isdigit()
+                ]
+                dir_nrs += [0]
+                _id = max(dir_nrs) + 1
+                _id = f'{prefix_}{_id}{_suffix}'
+        elif id_naming == 'time':
+            if i != 0:
+                time.sleep(1)
+            _id = datetime.datetime.now().strftime('%Y-%m-%d-%H-%M-%S')
+            _id = f'{prefix_}{_id}{_suffix}'
+        elif callable(id_naming):
+            _id = id_naming()
```

Comment:
Check if the ID is unique in the subfolder? This is very unlikely for our naming functions below, but this might not be the case for a user-supplied naming function.

Reply:
I don't get your point. The code later will check if the folder exist.
If it is the case, it will call again `id_naming` to get another name.
This is done up to 100 times, before the code complains, that it cannot find a new name.

Reply:
Ups, I didn't get that. Then it's fine!